feat: add (de)serialization customization

lsorber opened the pull request:
Adds the ability to specify a custom serialization/deserialization method to store/read the result of a dask computation.

Closes https://github.com/radix-ai/graphchain/issues/38.

multimeric commented on src/graphchain/core.py:
```diff
@@ -242,7 +254,7 @@ def time_to_result(self, memoize: bool = True) -> float:
     def cache_filename(self) -> str:
         """Filename of the cache file to load or store."""
-        return f"{self.hash}.joblib.lz4"
+        return f"{self.hash}.dat"
```
It might be good if this were customizable. For instance if I save to parquet it's because I want to be able to interrogate it with parquet reading tools, which will expect the `.parquet` extension.

multimeric replied:
For instance, you could add a new function to the optimizer which is a callable that returns the file extension for any given file.

lsorber replied:
Done. The serialize/deserialize function is now in control over the filename written to the filesystem.

multimeric commented on src/graphchain/core.py:
```diff
@@ -44,6 +45,11 @@ def layer_setitem(self: Layer, key: Hashable, value: Any) -> None:
 logger = logging.getLogger(__name__)
 
 
+def joblib_dump_lz4(obj: Any, fd: io.BytesIO) -> None:
```
Could this allow for a customizable compression, and then you also export a partial version of this which is the default? ie `joblib_dump_lz4 = partial(joblib_dump, compression="lz4")`

lsorber replied:
Done. The default implementation allows for customization as follows:
```
joblib_dump_gzip = partial(joblib_dump, compress="gzip", ext="joblib.gz", protocol=HIGHEST_PROTOCOL)
joblib_load_gzip = partial(joblib_load, ext="joblib.gz")
```

multimeric commented on src/graphchain/core.py:
```diff
@@ -349,32 +359,32 @@ def optimize(
     keys: Optional[Union[Hashable, Iterable[Hashable]]] = None,
     skip_keys: Optional[Container[Hashable]] = None,
     location: Union[str, fs.base.FS, CacheFS] = "./__graphchain_cache__",
+    serialize: Callable[[Any, io.BytesIO], None] = joblib_dump_lz4,
```
There may be issues with assuming any serializer can work with a file descriptor. For instance, dask's own `DataFrame.to_parquet` takes a **directory** and not a file descriptor, because it needs to be able to write out each partition to a separate file. Would it be okay to just pass the PyFilesystem path and the filesystem `cache_fs`, so that users can do what they want, with a docstring explaining that `self.cache_fs.open(fn, "rb")` is often what they will want to do?

lsorber replied:
Done. The serialize/deserialize API now gives you control over which files you write to the filesystem (if any). This is the default implementation that would come with graphchain:

```python
def joblib_dump(obj: Any, fs: fs.base.FS, key: str, **kwargs: Dict[str, Any]) -> None:
    """Store an object on a filesystem."""
    filename = f"{key}.joblib"
    if isinstance(kwargs.get("compress"), tuple) and len(kwargs["compress"]) == 2:
        filename = f"{filename}.{kwargs['compress'][0]}"  # type: ignore[index]
    elif isinstance(kwargs.get("compress"), str):
        filename = f"{filename}.{kwargs['compress']}"
    with fs.open(filename, "wb") as fid:
        joblib.dump(obj, fid, **kwargs)


def joblib_load(
    fs: fs.base.FS, key: str, filename_extension: str = "joblib", **kwargs: Dict[str, Any]
) -> Any:
    """Load an object from a filesystem."""
    filename = f"{key}.{filename_extension}"
    with fs.open(filename, "rb") as fid:
        return joblib.load(fid, **kwargs)

joblib_dump_lz4 = partial(joblib_dump, compress="lz4", protocol=HIGHEST_PROTOCOL)
joblib_load_lz4 = partial(joblib_load, filename_extension="joblib.lz4")
```

multimeric reviewed (commented):
Looks good! One thing I might add is that trying to use a parquet serializer as a test would be a good assessment of how well this approach works for directories.

lsorber commented on src/graphchain/core.py:
```diff
@@ -239,31 +264,24 @@ def time_to_result(self, memoize: bool = True) -> float:
     def cache_file_exists(self) -> bool:
         """Check if this ``CachedComputation``'s cache file exists."""
-        return self.cache_fs.exists(self.cache_filename)  # type: ignore[attr-defined,no-any-return]
+        return self.cache_fs.exists(f"{self.hash}.time.store")
```
This is a workaround for the fact that graphchain no longer knows which file(s) get written to the cache (if any). After calling the (user-supplied) `serialize` function, graphchain will also write the file `f"{self.hash}.time.store"` so this should work just fine and is probably even a bit more robust since we would only assume a cache file (or directory) is available after it has been written to completion.

Graphchain (and this feature) has been compatible with S3 from the beginning so yes, it should continue to work with S3.

lsorber commented:
> Looks good! One thing I might add is that trying to use a parquet serializer as a test would be a good assessment of how well this approach works for directories.

Thanks. Could you elaborate on what you mean with it working for directories? If the output of a dask computation is a DataFrame, I could serialise that to a parquet file, but that's just a single file.

multimeric commented:
Yes, I'm referring to `to_parquet`. That's my main motivation for requesting this feature.

multimeric commented:
Or I can add such a test, but I can't commit to doing so in the near future.

lsorber commented:
Added a test that demos using [dask.dataframe.to_parquet](https://docs.dask.org/en/stable/generated/dask.dataframe.to_parquet.html#dask.dataframe.to_parquet) as a custom serializer/deserializer for dask DataFrames, which also helped to catch a bug with dask dataframes.

multimeric commented:
Great!

jakirkham commented:
Would it make sense to include some kind of example either in the README or other docs?

multimeric commented:
Just copying the parquet test into the readme almost verbatim would probably work.

jakirkham commented:
Yes tests are a form of example. Though documentation can improve visibility. Up to you :)

multimeric commented:
Oh hmm, there aren't API docs are there. I guess that would be helpful now that the API has become a bit more complex. But that's maybe a job for another issue.

lsorber commented:
I'll update the README with an example, thanks for the suggestion! We used to have API documentation on Read the Docs, I might look into reinstating that.

@multimeric @jakirkham Would it make sense to use [dask.dataframe.to_parquet](https://docs.dask.org/en/stable/generated/dask.dataframe.to_parquet.html#dask.dataframe.to_parquet) by default on dask DataFrames (instead of joblib)? If yes, should we enable (fast) compression? And do we pick `fastparquet` or `pyarrow`?

multimeric commented:
No, I don't think it should be the default. If you have any Series that have the dtype of `object` then the parquet serializer will need a schema. So it's only applicable if the user keeps to primitive types and/or provides a parquet schema.

I'm not sure about the best compression or arrow library, but the dask default is snappy + pyarrow which you can probably copy.

lsorber commented:
Added a few additional improvements:
- [x] Relaxed and updated the dependencies
- [x] Improved the tests
- [x] Improved the docstrings
- [x] Improved and expanded the README with an example of the new (de)serialization customization

Any further suggestions or comments @jakirkham @multimeric?

jakirkham commented:
LGTM. Thanks for working on this 🙏

multimeric commented:
Yes, thank you indeed! I hope to make good use of this excellent improvement in the near future.

lsorber commented:
Merged! Thanks for your input @multimeric and @jakirkham!

jakirkham commented:
Thanks for working on this @lsorber! 🙏
